data/transactions/logic/opcodes.go

```diff
@@ -361,11 +361,117 @@ type typedList struct {
 	Effects string
 }
 
+// StackChangeExplanation explains the effect of an opcode over the stack.
+// An opcode may delete a few variables from stack, then add a few to stack.
+type StackChangeExplanation struct {
+	Deletions int
+	Additions int
+}
+
+type debugStackExplain func(*EvalContext) StackChangeExplanation
+
 // Proto describes the "stack behavior" of an opcode, what it pops as arguments
 // and pushes onto the stack as return values.
 type Proto struct {
 	Arg    typedList // what gets popped from the stack
 	Return typedList // what gets pushed to the stack
+
+	// Explain is the pointer to the function used in debugging process during simulation:
+	// - on default construction, Explain relies on Arg and Return count.
+	// - otherwise, we need to explicitly infer from EvalContext, by registering through explain function
+	Explain debugStackExplain
+}
+
+func (p Proto) stackExplain(e debugStackExplain) Proto {
+	p.Explain = e
+	return p
+}
+
+func defaultDebugExplain(argCount, retCount int) debugStackExplain {
+	return func(_ *EvalContext) StackChangeExplanation {
+		return StackChangeExplanation{Deletions: argCount, Additions: retCount}
+	}
+}
+
+// NextStackChange is a helper function that queries EvalContext for the coming stack change of the current PC.
+func (cx *EvalContext) NextStackChange() StackChangeExplanation {
+	return (opsByOpcode[cx.version][cx.program[cx.pc]].Explain)(cx)
+}
+
+func opReturnStackChange(cx *EvalContext) StackChangeExplanation {
+	return StackChangeExplanation{Deletions: len(cx.Stack), Additions: 1}
+}
+
+func opBuryStackChange(cx *EvalContext) StackChangeExplanation {
+	depth := int(cx.program[cx.pc+1])
+	return StackChangeExplanation{Deletions: depth + 1, Additions: depth}
+}
+
+func opPopNStackChange(cx *EvalContext) StackChangeExplanation {
+	n := int(cx.program[cx.pc+1])
+	return StackChangeExplanation{Deletions: n}
+}
+
+func opDupNStackChange(cx *EvalContext) StackChangeExplanation {
+	n := int(cx.program[cx.pc+1])
+	return StackChangeExplanation{Deletions: 1, Additions: n + 1}
+}
+
+func opDigStackChange(cx *EvalContext) StackChangeExplanation {
+	depth := int(cx.program[cx.pc+1])
+	return StackChangeExplanation{Deletions: depth + 1, Additions: depth + 2}
+}
```

Comment:
`dig` and `frame_dig` don't actually modify the lower part of the stack, so I think I'd prefer to view them as deleting 0 and adding 1 element (maybe they don't even need a special function then?)

Reply:
sure, both `dig` and `frame_dig` can be viewed as 0 del and 1 addition.

`frame_dig` has proto `:a`, so I just removed the special function; `dig` has proto `a:aa`, so I modified the stack change method to 0 del and 1 addition.

Reply:
Thanks for that change.

Out of curiosity, @jannotti do you know why `dig` has proto `a:aa`?

Reply:
Because it is guaranteed to fail if the stack is empty, so `a:aa` does a slightly better static check.

Reply:
I agree that, barring too much complexity, I'd like to see the trace claim a delete of 0, and a single add.

Reply:
This reminds me of a previous discussion with JJ, on why we want `a:aa` for frame dig: we are assuming something under there over the stack (at least 1), so the first `a` is there.

But in this case, we probably want to have 0 del and 1 add for both, so I will make the change to pin the stack explanation for both dig opcodes.